### ISSUE TITLE
MF_NON_INTERACTIVE does not seem to work at all

### ISSUE THREAD

**Issue by ezzatron** [make-files/adr#77]
Currently we manually set `MF_NON_INTERACTIVE` to `true` under CI, and try to detect it otherwise. But I'm pretty sure the detection is not working at all. After looking into it, I'm not even sure it's possible to detect an interactive shell under `make`. None of the suggestions I have seen hold up to actual testing (e.g. https://stackoverflow.com/questions/4251559/how-can-i-tell-if-a-makefile-is-being-run-from-an-interactive-shell).

Unless @jmalloc you have any ideas about how to actually detect non-interactivity under `make`, perhaps we should either:

- Stop passing "non-interactive" options to commands completely, and let the individual commands handle it; OR
- Always pass "non-interactive" options to commands; OR
- Always pass "non-interactive" options to commands under CI, but never outside of CI

The following Makefile:

```Makefile
DEFAULT_GOAL := test

export EXAMPLE_A := $(shell [ -t 0 ] || echo true)
export EXAMPLE_B := $(shell [[ $$- == *i* ]] && echo true)
export EXAMPLE_C := $(shell echo $$-)

test:
	echo "EXAMPLE_A=$(EXAMPLE_A)"
	echo "EXAMPLE_B=$(EXAMPLE_B)"
	echo "EXAMPLE_C=$(EXAMPLE_C)"
```

Gives the following output under various interactive and non-interactive circumstances:

```
$ make
echo "EXAMPLE_A="
EXAMPLE_A=
echo "EXAMPLE_B="
EXAMPLE_B=
echo "EXAMPLE_C=hBc"
EXAMPLE_C=hBc

$ bash -c make
echo "EXAMPLE_A="
EXAMPLE_A=
echo "EXAMPLE_B="
EXAMPLE_B=
echo "EXAMPLE_C=hBc"
EXAMPLE_C=hBc

$ make > /dev/stdout
echo "EXAMPLE_A="
EXAMPLE_A=
echo "EXAMPLE_B="
EXAMPLE_B=
echo "EXAMPLE_C=hBc"
EXAMPLE_C=hBc
```

**Comment by jmalloc** [comment]
Based on our discussion this morning it seems like our logic might be lacking there.

Here are some example expression that checks _both_ `STDIN` and `STDOUT` are TTYs and hence _likely_ to be interactive:

```Makefile
xxx:
	[[ -t 0 && -t 1 ]] && echo YES
```

Both TTYs:

```console
$ make
[[ -t 0 && -t 1 ]] && echo YES
YES
```

`STDIN` not a TTY:

```console
$ echo foo | make
[[ -t 0 && -t 1 ]] && echo YES
make: *** [xxx] Error 1
```

`STDOUT` not a TTY:

```console
$ make | cat
[[ -t 0 && -t 1 ]] && echo YES
make: *** [xxx] Error 1
```

It _might_ be the case that performing this test directly in the recipe like this behaves differently to the way another sub-shell would detect it's `STDIN`/`STDOUT` too.  If we can't get this working I would opt for:

> Always pass "non-interactive" options to commands;

**Comment by ezzatron** [comment]
> It _might_ be the case that performing this test directly in the recipe like this behaves differently to the way another sub-shell would detect it's `STDIN`/`STDOUT` too.

Yeah, I think that was part of it too. We do the detection inside a `$(shell)` call currently, which could change the results. I'll give it a test.

**Comment by ezzatron** [comment]
Confirmed that `$(shell)` changes the results. Given the following `Makefile`:

```Makefile
RESULT := $(shell [[ -t 0 && -t 1 ]] && echo interactive || echo non-interactive)

test:
	@echo Direct:
	@[[ -t 0 && -t 1 ]] && echo interactive || echo non-interactive
	@echo Shell call:
	@echo $(RESULT)
```

I get the following results:

```shell
$ make # both TTYs
Direct:
interactive
Shell call:
non-interactive

$ echo | make # STDIN not a TTY
Direct:
non-interactive
Shell call:
non-interactive

$ make | cat # STDOUT not a TTY
Direct:
non-interactive
Shell call:
non-interactive
```
